File: client/lib/features/products/data/models/product_request.dart
/// DTO for creating a product.
class CreateProductRequest {
  final String name;
  final String? description;
  final String? category;
  final int? shelfLifeDays;
  final double basePrice;
  final String? imageUrl;

  const CreateProductRequest({
    required this.name,
    this.description,
    this.category,
    this.shelfLifeDays,
    required this.basePrice,
    this.imageUrl,
  });

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{'name': name, 'basePrice': basePrice};

    if (description != null) json['description'] = description;
    if (category != null) json['category'] = category;
    if (shelfLifeDays != null) json['shelfLifeDays'] = shelfLifeDays;
    if (imageUrl != null) json['imageUrl'] = imageUrl;

    return json;
  }
}

/// DTO for updating a product.
class UpdateProductRequest {
  final String name;
  final String? description;
  final String? category;
  final int? shelfLifeDays;
  final double basePrice;
  final String? imageUrl;

  const UpdateProductRequest({
    required this.name,
    this.description,
    this.category,
    this.shelfLifeDays,
    required this.basePrice,
    this.imageUrl,
  });

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'description': description,
      'category': category,
      'shelfLifeDays': shelfLifeDays,
      'basePrice': basePrice,
      'imageUrl': imageUrl,
    };
  }
}

File: server/src/main/java/com/ahmadramadhan/mudahtitip/product/dto/ProductRequest.java
package com.ahmadramadhan.mudahtitip.product.dto;

import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import jakarta.validation.constraints.PositiveOrZero;
import jakarta.validation.constraints.Size;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;

/**
 * DTO for creating/updating a product.
 */
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class ProductRequest {

    @NotBlank(message = "Nama produk wajib diisi")
    @Size(min = 2, max = 100, message = "Nama produk harus antara 2-100 karakter")
    private String name;

    @Size(max = 500, message = "Deskripsi maksimal 500 karakter")
    private String description;

    @Size(max = 50, message = "Kategori maksimal 50 karakter")
    private String category;

    @Positive(message = "Masa simpan harus positif")
    private Integer shelfLifeDays;

    @NotNull(message = "Harga dasar wajib diisi")
    @Positive(message = "Harga harus positif")
    private BigDecimal basePrice;

    @NotNull(message = "Stok wajib diisi")
    @PositiveOrZero(message = "Stok tidak boleh negatif")
    private Integer stock;

    private String imageUrl;
}
